### ISSUE TITLE
autopkgtest failures with 243-rc1, systemd-random-seed.service timeout during boot

### ISSUE THREAD

**Issue by mbiebl** [systemd/systemd#13252]
I'm getting autopkgtest failures with 243-rc1 which I don't get with v242 or v241 (in Debian).

The failing tests are boot-and-services:
```
autopkgtest [22:09:27]: test boot-and-services: [-----------------------
none
Created symlink /etc/systemd/system/default.target → /lib/systemd/system/graphical.target.
bash: line 1: 12129 Killed                  /tmp/autopkgtest.9T1Cpl/build.sfX/src/debian/tests/boot-and-services 2> >(tee -a /tmp/autopkgtest.9T1Cpl/boot-and-servi)
autopkgtest [22:09:28]: test process requested reboot with marker boot1
qemu-system-x86_64: terminating on signal 15 from pid 9917 (/usr/bin/python3)
<VirtSubproc>: failure: timed out waiting for "login prompt on ttyS0"
autopkgtest [22:10:29]: ERROR: testbed failure: cannot send to testbed: [Errno 32] Broken pipe
```
and boot-smoke:
```
autopkgtest [22:36:07]: test boot-smoke: [-----------------------
reboot #0
bash: line 1: 14259 Killed                  /tmp/autopkgtest.2ObqjG/build.Aqs/src/debian/tests/boot-smoke 2> >(tee -a /tmp/autopkgtest.2ObqjG/boot-smoke-stderr >&2)
autopkgtest [22:36:08]: test process requested reboot with marker 1
qemu-system-x86_64: terminating on signal 15 from pid 13725 (/usr/bin/python3)
<VirtSubproc>: failure: timed out waiting for "login prompt on ttyS0"
autopkgtest [22:37:10]: ERROR: testbed failure: cannot send to testbed: [Errno 32] Broken pipe
```
I'm using qemu to run the tests and haven't further investigated this.
Since I can't reproduce the failure with v241 and v242, I thought it worthwile filing anyway.

**Comment by mbiebl** [comment]
I'm a bit at a loss how to debug this so I can provide further information.
@martinpitt , @ddstreet  any hints?

**Comment by mbiebl** [comment]
autopkgtest dbg log of failing boot-and-services: 
[log.txt](https://github.com/systemd/systemd/files/3463523/log.txt)

**Comment by keszybz** [comment]
Maybe #13251 ?

**Comment by ddstreet** [comment]
> I'm a bit at a loss how to debug this so I can provide further information.
> @martinpitt , @ddstreet any hints?

You should be able to use the autopkgtest -o param to save all the logs/artifacts in a local dir, does the upstream-stdout or -stderr have any more details on the failure?

**Comment by mbiebl** [comment]
@keszybz I tried with 10b1d42e579ae1ca6e022b61c3e5a44e1f3f2b80 but it didn't help

@ddstreet I don't see anything obvious in those files. That said, tarred up the `-o logs` directory 
[logs-fail.tar.gz](https://github.com/systemd/systemd/files/3464071/logs-fail.tar.gz)

**Comment by mbiebl** [comment]
git bisect to the rescue. This failure is caused by #13137 and more specifically, git bisect points at 26ded55709947d936634f1de0f43dcf88f594621 as the first faulty commit.

**Comment by evverx** [comment]
@mbiebl could you pass `--show-boot` to `autopkgtest` and paste the logs here? Could you also try doubling the timeout via `--boot-timeout=120` to see if it helps to make the test "pass"?

**Comment by evverx** [comment]
Just to clarify, `--show-boot` should be passed after `qemu`. That is, something like
```
autopkgtest ... -- qemu --show-boot ...
```
should suffice.

**Comment by mbiebl** [comment]
@evverx thanks for the hint

```
[  OK  ] Started Update UTMP about System Boot/Shutdown.
[  OK  ] Started Network Time Synchronization.
[  OK  ] Reached target System Time Set.
[  OK  ] Reached target System Time Synchronized.
[  OK  ] Started Helper to synchronize boot up for ifupdown.
         Starting Raise network interfaces...
[  OK  ] Started Raise network interfaces.
[     *] A start job is running for Load/Save Random Seed (45s /qemu-system-x86_64: terminating on signal 15 from pid 22417 (/usr/bin/python3)
[    **] A start job is running for Load/Save Random Seed (45s / 10min)
<VirtSubproc>: failure: timed out waiting for "login prompt on ttyS0"
autopkgtest [00:36:15]: ERROR: testbed failure: cannot send to testbed: [Errno 32] Broken pipe
```
That confirms that it's the systemd-random-seed related problem.
Bumping the timeout to 120s did not suffice, I had to bump it to 240s, the result was:
```
[  201.680358] random: crng init doneoad/Save Random Seed (3min 19s / 10min)
```
That's an aweful long time...

**Comment by yuwata** [comment]
@mbiebl Could you test #13263?

**Comment by mbiebl** [comment]
Adding this for completeness sake: the default for `--timeout-reboot` is 60s.

**Comment by mbiebl** [comment]
Unfortunately the above commit does not help.
Those tests still time out after 60ab2d1a4b3b927ad4536d0d8fb4609fb4053d62 has been applied

**Comment by mbiebl** [comment]
This is from an attempt with `--reboot-timeout=180`

```
Starting version 243-rc1-45-g46ff70c+
[    1.232495] piix4_smbus 0000:00:01.3: SMBus Host Controller at 0x700, revision 0
[    1.253520] Floppy drive(s): fd0 is 2.88M AMI BIOS
[    1.263198] SCSI subsystem initialized
[    1.272863] FDC 0 is a S82078B
[    1.282583] input: VirtualPS/2 VMware VMMouse as /devices/platform/i8042/serio1/input/input3
[    1.283993] input: VirtualPS/2 VMware VMMouse as /devices/platform/i8042/serio1/input/input2
[    1.304208] PCI Interrupt Link [LNKC] enabled at IRQ 11
[    1.312482] scsi host0: ata_piix
[    1.316479] scsi host1: ata_piix
[    1.316964] ata1: PATA max MWDMA2 cmd 0x1f0 ctl 0x3f6 bmdma 0xc0a0 irq 14
[    1.317856] ata2: PATA max MWDMA2 cmd 0x170 ctl 0x376 bmdma 0xc0a8 irq 15
[    1.325304] PCI Interrupt Link [LNKD] enabled at IRQ 10
[    1.339335] PCI Interrupt Link [LNKA] enabled at IRQ 10
[    1.343542] virtio_blk virtio2: [vda] 41943040 512-byte logical blocks (21.5 GB/20.0 GiB)
[    1.350646]  vda: vda1
[    1.478951] ata2.00: ATAPI: QEMU DVD-ROM, 2.5+, max UDMA/100
[    1.483663] scsi 1:0:0:0: CD-ROM            QEMU     QEMU DVD-ROM     2.5+ PQ: 0 ANSI: 5
[    1.522142] sr 1:0:0:0: [sr0] scsi3-mmc drive: 4x/4x cd/rw xa/form2 tray
[    1.523540] cdrom: Uniform CD-ROM driver Revision: 3.20
Begin: Loading essential drivers ... done.
Begin: Running /scripts/init-premount ... done.
Begin: Mounting root file system ... Begin: Running /scripts/local-top ... done.
Begin: Running /scripts/local-premount ... done.
Begin: Will now check root file system ... fsck from util-linux 2.34
[    1.736988] random: fast init done
[/sbin/fsck.ext4 (1) -- /dev/vda1] fsck.ext4 -a -C0 /dev/vda1 
/dev/vda1: clean, 43771/1310720 files, 609260/5242624 blocks
done.
[    1.749611] cryptd: max_cpu_qlen set to 1000
[    1.794566] EXT4-fs (vda1): mounted filesystem with ordered data mode. Opts: (null)
done.
Begin: Running /scripts/local-bottom ... done.
Begin: Running /scripts/init-bottom ... done.
[    1.847970] systemd[1]: Inserted module 'autofs4'
[    1.856364] systemd[1]: systemd 243-rc1-45-g46ff70c+ running in system mode. (+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT )
[    1.859371] systemd[1]: Detected virtualization kvm.
[    1.860053] systemd[1]: Detected architecture x86-64.

Welcome to Debian GNU/Linux bullseye/sid!

[    1.861930] systemd[1]: Set hostname to <host>.
[    1.962406] systemd[1]: /lib/systemd/system/dbus.socket:5: ListenStream= references a path below legacy directory /var/run/, updating /var/run/dbus/system_bus_s.
[    1.971816] systemd[1]: unit_file_find_fragment: getty@tty1.service has alias autovt@tty1.service
[    1.984329] random: systemd: uninitialized urandom read (16 bytes read)
[    1.985934] systemd[1]: Created slice system-getty.slice.
[  OK  ] Created slice system-getty.slice.
[    1.987781] random: systemd: uninitialized urandom read (16 bytes read)
[    1.988862] systemd[1]: Created slice system-serial\x2dgetty.slice.
[  OK  ] Created slice system-serial\x2dgetty.slice.
[    1.990901] random: systemd: uninitialized urandom read (16 bytes read)
[    1.991995] systemd[1]: Created slice User and Session Slice.
[  OK  ] Created slice User and Session Slice.
[    1.993874] systemd[1]: Started Dispatch Password Requests to Console Directory Watch.
[  OK  ] Started Dispatch Password …ts to Console Directory Watch.
[  OK  ] Started Forward Password R…uests to Wall Directory Watch.
[  OK  ] Set up automount Arbitrary…s File System Automount Point.
[  OK  ] Reached target Local Encrypted Volumes.
[  OK  ] Reached target User and Group Name Lookups.
[  OK  ] Reached target Paths.
[  OK  ] Reached target Remote File Systems.
[  OK  ] Reached target Slices.
[  OK  ] Reached target Swap.
[  OK  ] Listening on Syslog Socket.
[  OK  ] Listening on Process Core Dump Socket.
[  OK  ] Listening on fsck to fsckd communication Socket.
[  OK  ] Listening on initctl Compatibility Named Pipe.
[  OK  ] Listening on Journal Audit Socket.
[  OK  ] Listening on Journal Socket (/dev/log).
[  OK  ] Listening on Journal Socket.
[  OK  ] Listening on udev Control Socket.
[  OK  ] Listening on udev Kernel Socket.
         Mounting Huge Pages File System...
         Mounting POSIX Message Queue File System...
         Mounting Kernel Debug File System...
         Starting Create list of re…odes for the current kernel...
         Starting Journal Service...
         Starting Load Kernel Modules...
         Starting Remount Root and Kernel File Systems...
         Starting udev Coldplug all Devices...
[  OK  ] Mounted Huge Pages File System.
[  OK  ] Mounted POSIX Message Queue File System.
[  OK  ] Mounted Kernel Debug File System.
[  OK  ] Started Create list of req… nodes for the current kernel.
[    2.083820] EXT4-fs (vda1): re-mounted. Opts: errors=remount-ro
[  OK  ] Started Load Kernel Modules.
[  OK  ] Started Remount Root and Kernel File Systems.
         Starting Load/Save Random Seed...
         Starting Apply Kernel Variables...
         Starting Create System Users...
[  OK  ] Started Apply Kernel Variables.
[  OK  ] Started Create System Users.
         Starting Create Static Device Nodes in /dev...
[  OK  ] Started Create Static Device Nodes in /dev.
[  OK  ] Reached target Local File Systems (Pre).
[  OK  ] Reached target Local File Systems.
[  OK  ] Reached target Containers.
         Starting Load AppArmor profiles...
         Starting udev Kernel Device Manager...
[  OK  ] Started udev Coldplug all Devices.
         Starting Helper to synchronize boot up for ifupdown...
[    2.274837] audit: type=1400 audit(1565077587.652:2): apparmor="STATUS" operation="profile_load" profile="unconfined" name="/usr/sbin/haveged" pid=173 comm="app"
[    2.279822] audit: type=1400 audit(1565077587.660:3): apparmor="STATUS" operation="profile_load" profile="unconfined" name="nvidia_modprobe" pid=174 comm="appar"
[    2.281841] audit: type=1400 audit(1565077587.660:4): apparmor="STATUS" operation="profile_load" profile="unconfined" name="nvidia_modprobe//kmod" pid=174 comm="
[    2.291453] audit: type=1400 audit(1565077587.668:5): apparmor="STATUS" operation="profile_load" profile="unconfined" name="lsb_release" pid=175 comm="apparmor_"
[  OK  ] Started Load AppArmor profiles.
[  OK  ] Started udev Kernel Device Manager.
[    2.386766] input: Power Button as /devices/LNXSYSTM:00/LNXPWRBN:00/input/input4
[    2.387826] ACPI: Power Button [PWRF]
[    2.467333] parport_pc 00:04: reported by Plug and Play ACPI
[    2.468282] parport0: PC-style at 0x378, irq 7 [PCSPP,TRISTATE]
[    2.474850] input: PC Speaker as /devices/platform/pcspkr/input/input5
[    2.480582] sr 1:0:0:0: Attached scsi generic sg0 type 5
[    2.491783] 9pnet: Installing 9P2000 support
[  OK  ] Found device /dev/ttyS0.
[    2.708193] ppdev: user-space parallel port driver
[    2.803350] [drm] Found bochs VGA, ID 0xb0c0.
[    2.803956] [drm] Framebuffer size 16384 kB @ 0xfd000000, mmio @ 0xfebd0000.
[    2.816479] [TTM] Zone  kernel: Available graphics memory: 505524 kiB
[    2.817356] [TTM] Initializing pool allocator
[    2.817929] [TTM] Initializing DMA pool allocator
[    2.820586] fbcon: bochsdrmfb (fb0) is primary device
[    2.832556] Console: switching to colour frame buffer device 128x48
[    2.841370] bochs-drm 0000:00:02.0: fb0: bochsdrmfb frame buffer device
[    2.842282] [drm] Initialized bochs-drm 1.0.0 20130925 for 0000:00:02.0 on minor 0
[  OK  ] Started Helper to synchronize boot up for ifupdown.
         Starting Raise network interfaces...
[  OK  ] Started Raise network interfaces.
[  OK  ] Started Journal Service.
         Starting Flush Journal to Persistent Storage...
[    3.093569] systemd-journald[153]: Received client request to flush runtime journal.
[  OK  ] Started Flush Journal to Persistent Storage.
         Starting Create Volatile Files and Directories...
[  OK  ] Started Create Volatile Files and Directories.
         Starting Network Time Synchronization...
         Starting Update UTMP about System Boot/Shutdown...
[  OK  ] Started Update UTMP about System Boot/Shutdown.
[  OK  ] Started Network Time Synchronization.
[  OK  ] Reached target System Time Set.
[  OK  ] Reached target System Time Synchronized.
[   ***] A start job is running for Load/Savqemu-system-x86_64: terminating on signal 15 from pid 14109 (/usr/bin/python3)
[   ***] A start job is running for Load/Save Random Seed (2min 46s / 10min)
<VirtSubproc>: failure: timed out waiting for "login prompt on ttyS0"
autopkgtest [09:49:13]: ERROR: testbed failure: cannot send to testbed: [Errno 32] Broken pipe
```

**Comment by poettering** [comment]
@mbiebl, hmm, there shouldn't be any ordering dep on systemd-random-seed.service after that PR got merged. If you use "systemctl show" on the unit in those tests, what do you see?

**Comment by mbiebl** [comment]
Haha, this is hilarious.
```
# systemctl show systemd-random-seed.service -p Before -p After
Before=shutdown.target haveged.service systemd-boot-system-token.service
After=systemd-journald.socket system.slice -.mount systemd-remount-fs.service

# systemctl cat haveged.service     
# /lib/systemd/system/haveged.service
[Unit]
Description=Entropy daemon using the HAVEGE algorithm
Documentation=man:haveged(8) http://www.issihosts.com/haveged/
DefaultDependencies=no
ConditionVirtualization=!container
After=apparmor.service systemd-random-seed.service systemd-tmpfiles-setup.service
Before=sysinit.target shutdown.target

[Service]
EnvironmentFile=-/etc/default/haveged
ExecStart=/usr/sbin/haveged --Foreground --verbose=1 $DAEMON_ARGS
SuccessExitStatus=143
SecureBits=noroot-locked
NoNewPrivileges=yes
CapabilityBoundingSet=CAP_SYS_ADMIN
PrivateTmp=yes
PrivateDevices=yes
PrivateNetwork=yes
ProtectSystem=full
ProtectHome=yes

[Install]
WantedBy=default.target
```
To work around the issues with entropy starvation, it seems haveged is now installed by default in d-i [1] and apparently also autopkgtest [2].
And this apparently has now and adverse affect.
Somehow this whole entropy issue has been a recurring PITA over the last year(s) :-/


[1] https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=923675
[2] https://salsa.debian.org/ci-team/autopkgtest/blob/master/setup-commands/setup-testbed#L248

**Comment by mbiebl** [comment]
So if we have a service that hooks into `sysinit.target` and orders itself after `systemd-random-seed.service` this all fails spectacularly.

**Comment by mbiebl** [comment]
I scanned the Debian archive and haveged is currently the only service that uses `Before=sysinit.target` and `After=systemd-random-seed.service`

**Comment by poettering** [comment]
So, I doubt that haveged is really a great idea. if it was the kernel would do that stuff on its own...

I don#t understand why haveged wants to be ordered after the random seed stuff though. I don't think it should be... is there any rationale for that? it's not a consumer of entropy, is it? so why delay it needlessly?

why is haveged installed in those environments? Why not just do RNG passhtrough for autopkgtest? I mean, it runs its own qemu environments, no?

It would just require to do the equivalent of:

https://github.com/systemd/mkosi/pull/353

I am pretty sure this should be fixed in debian/autopkgtest, by either:

1. dropping the ordering dependency from haveged
2. adding the RNG passhtrough stuff to qemu (and then getting rid of haveged maybe)
3. accepting that boot takes a bit longer?

**Comment by poettering** [comment]
(and of course haveged and things are generally not really fixing any problems anyway, since they come too late: https://systemd.io/RANDOM_SEEDS#frequently-asked-questions)

**Comment by mbiebl** [comment]
@martinpitt I guess you can best answer why autopkgtest does install haveged.

d-i installs haveged as it "solves" the real world issue of boot stalling for minutes. Maybe it's not a good solution but apparently there isn't a better available atm? It feels like the blame is passed around here and maintainers trying to get something workable for our users.

The problem with "accepting that boot takes a bit longer?" is not, that it takes a bit longer, but that autopkgtest fails. And raising the timeout to 4min is not just a bit.

I'm also worried, that more services will start ordering themselves after systemd-random-seed.service (it's what you suggested yourself after all).

Regarding haveged, I'll ask the Debian maintainer, I can't answer this question.

**Comment by poettering** [comment]
> d-i installs haveged as it "solves" the real world issue of boot stalling for minutes. Maybe it's not a good solution but apparently there isn't a better available atm?

Well, the entropy has to be there, if code synchronously waits for it. It has to come from somewhere... That can be haveged (which is as mentioned a bit questionnable) or can be something else, but it needs to be there, it cannot be made of thin air...

Hmm, here's another idea: maybe those scripts should just provision a random seed on disk, and then tell systemd to credit it fully. With v243 this is possible like this:

```shell
dd if=/dev/urandom of=$ROOT/var/lib/systemd/random-seed bs=512 count=1
mkdir -p $ROOT/etc/systemd/system/systemd-random-seed.service.d
cat >  $ROOT/etc/systemd/system/systemd-random-seed.service.d/50-credit-entropy.conf <<EOF
[Service]
Environment=SYSTEMD_RANDOM_SEED_CREDIT=force
```

(Not tested, but you get the idea. For background see: https://systemd.io/RANDOM_SEEDS)

(this fix is only good in very specific cases where you know the image is replicated anywhere, i.e. it's fine for testing, but it should not be considered a good fix for the general case, as it means every instance of the machine comes up with the same pool).

It appears to me this is much better than doing haveged. That said it fixes the problem so that haveged won't do any good but also not any bad)

**Comment by martinpitt** [comment]
@mbiebl: haveged was introduced [in 2014](https://salsa.debian.org/ci-team/autopkgtest/commit/ffb46d6e0a6581cf1775) as part of launching cloud instances. Unfortunately that was copied from a separate project that now doesn't exist any more, so the precise rationale of it is lost in the mists of time. We did have tests or packages that required entropy at startup, which were then stalled for  minutes. E. g. in 2016 we got that "funny" [python 3.5](https://bugs.python.org/issue26839) change that blocked the boot again.

Indeed at that time that was the generally accepted and working solution for "these are test machines, I don't care about good random numbers, I just need them to boot, d****it!". Since then there are better tools to seed entropy into VMs. I don't know how much they found their way into OpenStack deployments yet.

However, upstream CI (the Ubuntu cloud runner) seems fine with master. You are saying that this only affects your local runs in QEMU, not upstream master? Maybe the reason is actually that systemd-random-seed.service works in the cloud, but previously just timed out/failed after 30 seconds, but after 
https://github.com/systemd/systemd/commit/26ded55709947d936634f1de0f43dcf88f594621 it now causes the timeout? Do you get any hangs if you remove haveged from your qemu?

I'm fine with trying to kick out haveged from the autopkgtest setup script and see how it goes. @iainlane, @xnox , would that be okay with you? I. e. this should happen at a time when you can watch for oddities/hangs in the infra, not in the middle of holidays.

**Comment by keszybz** [comment]
OK, let's close this. Please reopen if the issue still happens without haveged.

**Comment by mbiebl** [comment]
So we now just break systems for users who have haveged installed?

**Comment by keszybz** [comment]
`haveged.service` sets `DefaultDependencies=no` and orders itself before `sysinit.target`. That is OK, but it's the responsibility of the service to make sure it doesn't break the boot. Plugging into the early boot before `sysinit.target` is complicated and should not be attempted lightly. In this case `haveged.service` simply needs to be updated to match the changes in systemd.

**Comment by mbiebl** [comment]
Let's be clear, it's a change in systemd which breaks the boot/haveged, not the other way around.

That said what would be that change in haveged? How can haveged ensure to work with pre-243 and post-243?

**Comment by keszybz** [comment]
Yes, of course. But haveged.service plugs very deep into systemd internals, and when those internals change, anyone who plugs into those internals needs to be updated. Sorry, but you can't add orderings on various early boot services and expect no change over time.

**Comment by mbiebl** [comment]
@nbraud ^

**Comment by poettering** [comment]
Yes, we changed behaviour here. But as @keszybz said, during the early boot phase this is something we take more liberty to do than for the late boot phase. I am really curious though why haveged even wants to be ordered after the random seed service though. There must be some rationale behind that choice of ordering, and I am really curious what it is?

**Comment by ddstreet** [comment]
Looking upstream, the haveged-provided service files don't appear to be problematic (i.e. they don't have DefaultDependencies=no nor do they have Before=sysinit.target):
https://github.com/jirka-h/haveged/tree/1.9.4/init.d

Looking at the debian service file, it seems this commit was what added the early-boot configuration, based on what opensuse had done:
https://salsa.debian.org/debian/haveged/commit/6e04e09636e42deb0255c83d3d57f8d6e342dedb

This appears to be where opensuse changed the service file:
https://build.opensuse.org/request/show/246784

Unfortunately the referenced opensuse bug appears to be private:
https://bugzilla.opensuse.org/show_bug.cgi?id=892096

For Redhat/Fedora, I couldn't find any linkable online package source, but I looked at the service file installed in fedora fc30 and rhel el7, and neither of them have this problem (i.e. they do not include Before=sysinit.service nor DefaultDependencies=no).

Maybe it's only Suse and Debian/Ubuntu that include problematic haveged service files?

**Comment by keszybz** [comment]
@ddstreet in Fedora, all sources are available as `https://src.fedoraproject.org/rpms/<name>/`. In this case, https://src.fedoraproject.org/rpms/haveged/tree/master. The spec lists where the tarball was downloaded from. The file in question is https://src.fedoraproject.org/lookaside/pkgs/haveged/haveged.service/bd5446be4ab272f14737a269cdb5a78c/haveged.service. (Pfff, that's all non-obvious. I wish Fedora had this all in one place.).

**Comment by ddstreet** [comment]
I opened debian MP for haveged to remove the After=systemd-random-seed.service from its service file:
https://salsa.debian.org/debian/haveged/merge_requests/1

**Comment by mbiebl** [comment]
FYI: https://salsa.debian.org/ci-team/autopkgtest/merge_requests/60

**Comment by iainlane** [comment]
I'm talking to our IS team about whether we can use virtio-rng for the cloud instances we use to run these tests in the data centre. That's a reasonably normal Openstack cloud, and it [does seem like that is possible](https://wiki.openstack.org/wiki/LibvirtVirtioRng), so hopefully we'll be OK there and then we can stop using haveged.

**Comment by LeSpocky** [comment]
> So, I doubt that haveged is really a great idea. if it was the kernel would do that stuff on its own...

There are patches pending by Stephan Müller, which would rework the random subsystem in the kernel _and_ add another entropy source for virtual or embedded systems. Those devices suffer from low entropy in general.

> I don#t understand why haveged wants to be ordered after the random seed stuff though. I don't think it should be... is there any rationale for that? it's not a consumer of entropy, is it? so why delay it needlessly?

That makes no sense to me. haveged should be started as early as possible.

> why is haveged installed in those environments? Why not just do RNG passhtrough for autopkgtest? I mean, it runs its own qemu environments, no?

You may be able to pass some rng stuff from the host to a virtual environment, but that's not possible for an embedded system. Think of a device without HID like keyboard or mouse and maybe even no network. Without haveged or jitterentropy-rng you can literally wait days until the kernel gathered enough entropy and says "crng ready".

>     1. dropping the ordering dependency from haveged

It makes certainly no sense to start haveged after the crng is ready, it is needed before.

>     2. adding the RNG passhtrough stuff to qemu (and then getting rid of haveged maybe)

This probably makes sense, but is not applicable to embedded systems.

>     3. accepting that boot takes a bit longer?

Nope. Waiting for the crng might still take several seconds, although haveged is already started. That is quite long on an embedded system optimized for boot time.
